File: lib/tasks/determinism.rb
# frozen_string_literal: true

require "securerandom"
require "socket"
require "time"
require_relative "timezones"

module Tasks
  # The one place that turns harness *pins* into the injectable values the
  # library already accepts. It exists for the Go-port conformance harness
  # (`porting/`), where two runs of the same command must produce byte-identical
  # files, and for nothing else.
  #
  # Design rules, deliberately narrow:
  #
  #   * This module is an **adapter-boundary** concern, like Config. Only
  #     `bin/tasks`, `bin/tasks-api`, `bin/tasks-tui`, and Config itself read it.
  #     Store, Journal, Application and the domain stay env-free and keep taking
  #     plain injected values.
  #   * Every accessor returns `nil` when its pin is unset, and every caller
  #     falls back to the constructor default it used before. Unpinned behavior
  #     is therefore bit-for-bit what it was.
  #   * It configures nothing else. It is not a second config system: it cannot
  #     select files, change formatting, or alter semantics. If a pin would do
  #     any of those, it does not belong here.
  #
  # Pins are documented in `porting/specs/determinism.md`; that file and this
  # one must be changed together.
  module Determinism
    # Instant every clock reads. RFC3339 / ISO8601 with an offset (e.g.
    # "2026-03-14T15:09:26Z"). Unset: the real clock.
    NOW = "TASKS_PIN_NOW"

    # Task-id mint sequence. Comma-separated tokens; each is either eight hex
    # characters or the literal "seq" (equivalent to "00000000"). Ids are minted
    # in order; after the last token the sequence continues by incrementing it as
    # a 32-bit counter. Unset: SecureRandom.
    IDS = "TASKS_PIN_IDS"

    # The journal's coalescing scope — a per-process random token that is
    # persisted into journal index.json and therefore observable in journal
    # bytes. Unset: SecureRandom.hex(16).
    COALESCE_SCOPE = "TASKS_PIN_COALESCE_SCOPE"

    # The per-operation coalescing KEY minted for each delegation operation
    # (delegate / undelegate / claim / release). Like the scope, it is persisted
    # into journal index.json and is therefore observable in journal bytes —
    # unlike the scope, a fresh one is minted per operation rather than per
    # process, so it needs a sequence rather than a single value. Same spelling
    # rules as IDS but sixteen hex characters wide. Unset: SecureRandom.hex(8).
    DELEGATION_KEYS = "TASKS_PIN_DELEGATION_KEYS"

    # Host name used for host-context selection in Config. Unset:
    # Socket.gethostname. (The *device* half of update stamps has its own
    # long-standing setting, TASKS_DEVICE; this module does not duplicate it.)
    HOSTNAME = "TASKS_PIN_HOSTNAME"

    # Terminal geometry for full-screen clients. Standard POSIX names, honored
    # only when the real terminal size would otherwise be queried. Unset: the
    # tty's own winsize, then 24x80.
    LINES = "LINES"
    COLUMNS = "COLUMNS"

    # Pre-existing settings the harness also pins; listed so `report` can show
    # the complete applied set without inventing new names for them.
    DEVICE = "TASKS_DEVICE"
    TZ = "TZ"
    LANG = "LANG"
    LC_ALL = "LC_ALL"

    KEYS = [NOW, IDS, COALESCE_SCOPE, DELEGATION_KEYS, HOSTNAME, LINES, COLUMNS,
            DEVICE, TZ, LANG, LC_ALL].freeze

    # A monotonic, reproducible hex mint. Stateful on purpose: one CLI
    # invocation can perform several mutations, and each must draw the *next*
    # token rather than repeat the first. Store's existing collision loop still
    # applies to ids, so an id already present in the store or archive is
    # skipped exactly as before.
    #
    # `width` is the token length in hex characters — 8 for task/section ids,
    # 16 for delegation coalescing keys — and `name` is only used in messages,
    # so a malformed pin names the pin the operator actually set.
    class IdSequence
      SPAN = 1 << 32

      attr_reader :width

      def initialize(spec, width: 8, name: IDS)
        @width = Integer(width)
        @name = name
        @token = /\A[0-9a-f]{#{@width}}\z/
        @span = 16**@width
        @queue = parse(spec)
        raise ArgumentError, "#{@name} is empty" if @queue.empty?

        @counter = nil
      end

      def call
        if (token = @queue.shift)
          @counter = token.to_i(16)
          return token
        end
        @counter = (@counter + 1) % @span
        format("%0#{@width}x", @counter)
      end

      private

      def parse(spec)
        spec.to_s.split(",").map do |raw|
          token = raw.strip.downcase
          token = "0" * @width if token == "seq"
          unless @token.match?(token)
            raise ArgumentError,
                  "#{@name} token must be #{@width} hex characters or \"seq\", got #{raw.inspect}"
          end

          token
        end
      end
    end

    module_function

    # Frozen UTC instant, or nil. Raises on an unparseable pin: a harness that
    # silently fell back to the wall clock would produce a green run that proves
    # nothing.
    def now(env: ENV)
      raw = env[NOW].to_s.strip
      return nil if raw.empty?

      Time.iso8601(raw).utc.freeze
    rescue ArgumentError
      raise ArgumentError, "#{NOW} must be an ISO8601 instant with an offset, got #{raw.inspect}"
    end

    # Clock lambda in the shape Store/StoreFactory/TemporalContext already take.
    def clock(env: ENV)
      instant = now(env: env)
      instant && -> { instant }
    end

    # Process-wide id mint. Memoized because `bin/tasks` builds more than one
    # Store per invocation and they must share one sequence.
    def id_source(env: ENV)
      spec = env[IDS].to_s.strip
      return nil if spec.empty?
      return @id_source if defined?(@id_source) && @id_source_spec == spec

      @id_source_spec = spec
      @id_source = IdSequence.new(spec)
    end

    def coalesce_scope(env: ENV)
      value = env[COALESCE_SCOPE].to_s.strip
      value.empty? ? nil : value
    end

    # Process-wide mint for delegation coalescing keys. Memoized for the same
    # reason id_source is: one invocation may build more than one Application
    # and they must draw from one sequence rather than each restart it.
    def delegation_key_source(env: ENV)
      spec = env[DELEGATION_KEYS].to_s.strip
      return nil if spec.empty?
      return @delegation_key_source if defined?(@delegation_key_source) && @delegation_key_spec == spec

      @delegation_key_spec = spec
      @delegation_key_source = IdSequence.new(spec, width: 16, name: DELEGATION_KEYS)
    end

    # Callable, never nil: Config.resolve wants a hostname provider and the
    # unpinned provider is exactly the one it defaulted to before.
    def hostname(env: ENV)
      pinned = env[HOSTNAME].to_s.strip
      pinned.empty? ? -> { Socket.gethostname } : -> { pinned }
    end

    # [rows, columns] when both are pinned to positive integers, else nil so the
    # caller queries the real terminal.
    def winsize(env: ENV)
      rows = positive_integer(env[LINES])
      columns = positive_integer(env[COLUMNS])
      rows && columns ? [rows, columns] : nil
    end

    def positive_integer(value)
      number = Integer(value.to_s.strip, 10)
      number.positive? ? number : nil
    rescue ArgumentError, TypeError
      nil
    end

    # Everything an observation should record about how this process was pinned,
    # plus the host facts a pin cannot control but a comparison must know. Keys
    # with unset pins are present with a null value: "no pin was applied" is a
    # fact worth recording, and omission would make it indistinguishable from a
    # harness that forgot to look.
    def report(env: ENV)
      {
        "pins" => KEYS.to_h { |key| [key, env[key].nil? ? nil : env[key].to_s] },
        "tzdb_version" => Timezones.tzdb_version,
        "ruby_version" => RUBY_VERSION,
        "ruby_platform" => RUBY_PLATFORM,
        "default_external" => Encoding.default_external.to_s,
        "default_internal" => Encoding.default_internal&.to_s
      }
    end

    # Test-only: drop the memoized sequences.
    def reset!
      remove_instance_variable(:@id_source) if defined?(@id_source)
      remove_instance_variable(:@id_source_spec) if defined?(@id_source_spec)
      remove_instance_variable(:@delegation_key_source) if defined?(@delegation_key_source)
      remove_instance_variable(:@delegation_key_spec) if defined?(@delegation_key_spec)
      nil
    end
  end
end

File: internal/determinism/determinism.go
// Package determinism turns harness pins into the injectable values the read
// path already accepts. It is the Go counterpart of lib/tasks/determinism.rb
// and exists for the porting conformance harness and for nothing else.
//
// The design rules are the Ruby module's, unchanged: every accessor reports
// "unset" rather than substituting a value of its own, and it configures
// nothing that is not already a seam.
package determinism

import (
	"fmt"
	"math/big"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

// The pinned names, spelled exactly as lib/tasks/determinism.rb spells them.
const (
	NameNow               = "TASKS_PIN_NOW"
	NameIDs               = "TASKS_PIN_IDS"
	NameCoalesceScope     = "TASKS_PIN_COALESCE_SCOPE"
	NameDelegationKeys    = "TASKS_PIN_DELEGATION_KEYS"
	NameHostname          = "TASKS_PIN_HOSTNAME"
	NameLines             = "LINES"
	NameColumns           = "COLUMNS"
	NameTestTodaySequence = "TASKS_TEST_TODAY_SEQUENCE"
	NameDevice            = "TASKS_DEVICE"
	NameTZ                = "TZ"
	NameLang              = "LANG"
	NameLCAll             = "LC_ALL"
)

// Keys is Determinism::KEYS: every name `report` accounts for, in the Ruby
// module's own order.
var Keys = []string{
	NameNow, NameIDs, NameCoalesceScope, NameDelegationKeys, NameHostname,
	NameLines, NameColumns, NameTestTodaySequence, NameDevice, NameTZ,
	NameLang, NameLCAll,
}

// Env is a process environment. A name absent from the map is unset; a name
// present with an empty value is set-and-empty, which several pins treat
// differently from unset — TASKS_TEST_TODAY_SEQUENCE keys off presence alone,
// exactly as `if ENV[name]` does under Ruby's truthiness.
type Env map[string]string

// OSEnv snapshots the real environment.
func OSEnv() Env {
	env := Env{}
	for _, entry := range os.Environ() {
		if index := strings.IndexByte(entry, '='); index > 0 {
			env[entry[:index]] = entry[index+1:]
		}
	}
	return env
}

// Lookup reports the value and whether the name is set at all.
func (e Env) Lookup(name string) (string, bool) {
	value, present := e[name]
	return value, present
}

// Get is the `ENV[name].to_s` spelling: unset reads as the empty string.
func (e Env) Get(name string) string { return e[name] }

// Requested is the probe's `requested` helper: the raw value, or "" when the
// name is unset or blank. A pin nobody asked for is not a pin that was dropped.
func (e Env) Requested(name string) (string, bool) {
	raw, present := e[name]
	if !present || strings.TrimSpace(raw) == "" {
		return "", false
	}
	return raw, true
}

// isoInstant is Ruby's Time.xmlschema pattern (Time.iso8601 is its alias),
// including the case-insensitivity and the surrounding-whitespace tolerance.
var isoInstant = regexp.MustCompile(`\A\s*(-?\d+)-(\d\d)-(\d\d)[Tt](\d\d):(\d\d):(\d\d)(\.\d+)?([Zz]|[+-]\d\d(?::?\d\d)?)?\s*\z`)

// Now is the frozen UTC instant TASKS_PIN_NOW names, or ok=false when the pin
// is unset. An unparseable pin is an error rather than a silent fall back to
// the wall clock: a harness that fell back would produce a green run that
// proves nothing.
func Now(env Env) (time.Time, bool, error) {
	raw := strings.TrimSpace(env.Get(NameNow))
	if raw == "" {
		return time.Time{}, false, nil
	}
	instant, err := parseISO8601(raw)
	if err != nil {
		return time.Time{}, false, fmt.Errorf("%s must be an ISO8601 instant with an offset, got %q", NameNow, raw)
	}
	return instant.UTC(), true, nil
}

// NowForAdapter resolves the whole clock precedence in one place:
// TASKS_PIN_NOW dominates, then TASKS_TEST_TODAY_SEQUENCE (noon UTC of the
// current date), then the wall clock. Presence, not non-emptiness, triggers
// step 2 — the Ruby seam it mirrors reads `if ENV[name]`.
func NowForAdapter(env Env) (time.Time, error) {
	pinned, ok, err := Now(env)
	if err != nil {
		return time.Time{}, err
	}
	if ok {
		return pinned, nil
	}
	if _, present := env.Lookup(NameTestTodaySequence); !present {
		return time.Now().UTC(), nil
	}
	today := time.Now()
	return time.Date(today.Year(), today.Month(), today.Day(), 12, 0, 0, 0, time.UTC), nil
}

// Clock is the pinned instant in the shape the store takes, or nil when unset.
func Clock(env Env) func() time.Time {
	instant, ok, err := Now(env)
	if err != nil || !ok {
		return nil
	}
	return func() time.Time { return instant }
}

// CoalesceScope is the journal's per-process coalescing scope, or ok=false.
func CoalesceScope(env Env) (string, bool) {
	value := strings.TrimSpace(env.Get(NameCoalesceScope))
	return value, value != ""
}

// Hostname is Config's host provider: never nil, and the unpinned provider is
// the one Config defaulted to before.
func Hostname(env Env) func() string {
	pinned := strings.TrimSpace(env.Get(NameHostname))
	if pinned == "" {
		return func() string {
			name, err := os.Hostname()
			if err != nil {
				return ""
			}
			return name
		}
	}
	return func() string { return pinned }
}

// Winsize is [rows, columns] when both are pinned to positive integers, so the
// caller falls back to querying a real terminal only when they are not.
func Winsize(env Env) (rows, columns int, ok bool) {
	rows, rowsOK := positiveInteger(env.Get(NameLines))
	columns, columnsOK := positiveInteger(env.Get(NameColumns))
	if rowsOK && columnsOK {
		return rows, columns, true
	}
	return 0, 0, false
}

func positiveInteger(value string) (int, bool) {
	// Ruby's Integer(value, 10) accepts a leading sign and underscores and
	// rejects everything else, including an empty string.
	number, err := strconv.Atoi(strings.ReplaceAll(strings.TrimSpace(value), "_", ""))
	if err != nil || number <= 0 {
		return 0, false
	}
	return number, true
}

// shared holds the process-wide mints. Ruby memoizes these inside the module
// for a reason that is not cosmetic: ONE invocation can build several stores
// (a CLI store and an application factory), and they must draw from one
// sequence rather than each restarting it at the first pinned token — two
// records would otherwise be minted the same id and the store's collision loop
// would silently renumber the second.
var shared struct {
	sync.Mutex
	idSpec         string
	idSource       *IDSequence
	delegationSpec string
	delegationKeys *IDSequence
}

// SharedIDSource is IDSource memoized per spec: the same env yields the same
// sequence, and changing the pin rebuilds it. Adapters that mint ids more than
// once in a process should use this rather than IDSource.
func SharedIDSource(env Env) (*IDSequence, error) {
	spec := strings.TrimSpace(env.Get(NameIDs))
	if spec == "" {
		return nil, nil
	}
	shared.Lock()
	defer shared.Unlock()
	if shared.idSource != nil && shared.idSpec == spec {
		return shared.idSource, nil
	}
	sequence, err := NewIDSequence(spec, 8, NameIDs)
	if err != nil {
		return nil, err
	}
	shared.idSpec, shared.idSource = spec, sequence
	return sequence, nil
}

// SharedDelegationKeySource is DelegationKeySource memoized per spec, for the
// same reason: one invocation may build more than one application.
func SharedDelegationKeySource(env Env) (*IDSequence, error) {
	spec := strings.TrimSpace(env.Get(NameDelegationKeys))
	if spec == "" {
		return nil, nil
	}
	shared.Lock()
	defer shared.Unlock()
	if shared.delegationKeys != nil && shared.delegationSpec == spec {
		return shared.delegationKeys, nil
	}
	sequence, err := NewIDSequence(spec, 16, NameDelegationKeys)
	if err != nil {
		return nil, err
	}
	shared.delegationSpec, shared.delegationKeys = spec, sequence
	return sequence, nil
}

// Reset drops the memoized sequences. Test-only, exactly as Determinism.reset!
// is: a test that pinned one spec must not leak its position into the next.
func Reset() {
	shared.Lock()
	defer shared.Unlock()
	shared.idSpec, shared.idSource = "", nil
	shared.delegationSpec, shared.delegationKeys = "", nil
}

// IDSource is a fresh id mint, or nil when TASKS_PIN_IDS is unset.
func IDSource(env Env) (*IDSequence, error) {
	spec := strings.TrimSpace(env.Get(NameIDs))
	if spec == "" {
		return nil, nil
	}
	return NewIDSequence(spec, 8, NameIDs)
}

// DelegationKeySource is the process-wide mint for delegation coalescing keys,
// or nil when TASKS_PIN_DELEGATION_KEYS is unset.
func DelegationKeySource(env Env) (*IDSequence, error) {
	spec := strings.TrimSpace(env.Get(NameDelegationKeys))
	if spec == "" {
		return nil, nil
	}
	return NewIDSequence(spec, 16, NameDelegationKeys)
}

// IDSequence is a monotonic, reproducible hex mint. Stateful on purpose: one
// invocation can perform several mutations and each must draw the next token.
type IDSequence struct {
	// A shared sequence can be drawn from more than one goroutine (the store
	// and the journal both mint), and the draw mutates the queue and counter.
	mu      sync.Mutex
	width   int
	name    string
	queue   []string
	counter *big.Int
	span    *big.Int
}

var hexToken = regexp.MustCompile(`\A[0-9a-f]+\z`)

// NewIDSequence parses a pin spec: comma-separated tokens, each `width` hex
// characters or the literal "seq".
func NewIDSequence(spec string, width int, name string) (*IDSequence, error) {
	queue := []string{}
	// Ruby's String#split drops TRAILING empty fields, so ",," yields no tokens
	// while "a,,b" yields three. Mirrored so an empty spec reports "is empty"
	// rather than a token diagnostic about a field Ruby never produced.
	fields := strings.Split(spec, ",")
	for len(fields) > 0 && fields[len(fields)-1] == "" {
		fields = fields[:len(fields)-1]
	}
	for _, raw := range fields {
		token := strings.ToLower(strings.TrimSpace(raw))
		if token == "seq" {
			token = strings.Repeat("0", width)
		}
		if len(token) != width || !hexToken.MatchString(token) {
			return nil, fmt.Errorf("%s token must be %d hex characters or \"seq\", got %q", name, width, raw)
		}
		queue = append(queue, token)
	}

	if len(queue) == 0 {
		return nil, fmt.Errorf("%s is empty", name)
	}
	span := new(big.Int).Exp(big.NewInt(16), big.NewInt(int64(width)), nil)
	return &IDSequence{width: width, name: name, queue: queue, span: span}, nil
}

// Width is the token length in hex characters.
func (s *IDSequence) Width() int { return s.width }

// Call draws the next token: the queued ones in order, then the last one
// incremented as a counter of the same width.
func (s *IDSequence) Call() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.queue) > 0 {
		token := s.queue[0]
		s.queue = s.queue[1:]
		counter, _ := new(big.Int).SetString(token, 16)
		s.counter = counter
		return token
	}
	s.counter = new(big.Int).Mod(new(big.Int).Add(s.counter, big.NewInt(1)), s.span)
	return fmt.Sprintf("%0*s", s.width, s.counter.Text(16))
}

func parseISO8601(raw string) (time.Time, error) {
	match := isoInstant.FindStringSubmatch(raw)
	if match == nil {
		return time.Time{}, fmt.Errorf("not an ISO8601 instant")
	}
	year, err := strconv.Atoi(match[1])
	if err != nil {
		return time.Time{}, err
	}
	month, _ := strconv.Atoi(match[2])
	day, _ := strconv.Atoi(match[3])
	hour, _ := strconv.Atoi(match[4])
	minute, _ := strconv.Atoi(match[5])
	second, _ := strconv.Atoi(match[6])
	nanos := 0
	if match[7] != "" {
		fraction := match[7][1:]
		for len(fraction) < 9 {
			fraction += "0"
		}
		nanos, _ = strconv.Atoi(fraction[:9])
	}
	location := time.Local
	switch offset := match[8]; {
	case offset == "":
		// No offset: Ruby's Time.xmlschema falls back to local time.
	case offset == "Z" || offset == "z":
		location = time.UTC
	default:
		sign := 1
		if offset[0] == '-' {
			sign = -1
		}
		digits := strings.ReplaceAll(offset[1:], ":", "")
		hours, _ := strconv.Atoi(digits[:2])
		minutes := 0
		if len(digits) > 2 {
			minutes, _ = strconv.Atoi(digits[2:])
		}
		location = time.FixedZone("", sign*(hours*3600+minutes*60))
	}
	return time.Date(year, time.Month(month), day, hour, minute, second, nanos, location), nil
}
